File: rsocket-service/src/main/java/com/example/service/ServiceApplication.java
package com.example.service;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.RequiredArgsConstructor;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.data.annotation.Id;
import org.springframework.data.repository.reactive.ReactiveCrudRepository;
import org.springframework.messaging.handler.annotation.MessageMapping;
import org.springframework.messaging.rsocket.RSocketStrategies;
import org.springframework.messaging.rsocket.annotation.support.RSocketMessageHandler;
import org.springframework.security.config.Customizer;
import org.springframework.security.config.annotation.rsocket.EnableRSocketSecurity;
import org.springframework.security.config.annotation.rsocket.RSocketSecurity;
import org.springframework.security.config.web.server.ServerHttpSecurity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.core.userdetails.MapReactiveUserDetailsService;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.messaging.handler.invocation.reactive.AuthenticationPrincipalArgumentResolver;
import org.springframework.security.rsocket.core.PayloadSocketAcceptorInterceptor;
import org.springframework.security.web.server.SecurityWebFilterChain;
import org.springframework.stereotype.Controller;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.ServerResponse;
import org.springframework.web.reactive.handler.SimpleUrlHandlerMapping;
import org.springframework.web.reactive.socket.WebSocketHandler;
import org.springframework.web.reactive.socket.WebSocketMessage;
import org.springframework.web.reactive.socket.server.support.WebSocketHandlerAdapter;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.security.Principal;
import java.time.Duration;
import java.time.Instant;
import java.util.Map;
import java.util.stream.Stream;

import static org.springframework.web.reactive.function.server.RouterFunctions.route;
import static org.springframework.web.reactive.function.server.ServerResponse.ok;

@EnableRSocketSecurity
@SpringBootApplication
public class ServiceApplication {

    public static void main(String[] args) {
        SpringApplication.run(ServiceApplication.class, args);
    }

    @Bean
    RouterFunction<ServerResponse> routes(ReservationRepository rr) {
        return route()
                .GET("/reservations", request -> ok().body(rr.findAll(), Reservation.class))
                .build();
    }

    @Bean
    SecurityWebFilterChain httpAuthorization(ServerHttpSecurity httpSecurity) {
        return httpSecurity
                .httpBasic(Customizer.withDefaults())
                .csrf(ServerHttpSecurity.CsrfSpec::disable)
                .authorizeExchange(ae -> ae.anyExchange().authenticated())
                .build();
    }

    @Bean
    RSocketMessageHandler rSocketMessageHandler (RSocketStrategies strategies) {
        var mh = new RSocketMessageHandler();
        mh.getArgumentResolverConfigurer().addCustomResolver(new AuthenticationPrincipalArgumentResolver());
        mh.setRSocketStrategies(strategies);
        return mh;
    }

    @Bean
    PayloadSocketAcceptorInterceptor rsocketAuthorization(RSocketSecurity security) {
        return security
                .simpleAuthentication(Customizer.withDefaults())
                .authorizePayload(ae -> ae.anyExchange().authenticated())
                .build();
    }

    @Bean
    MapReactiveUserDetailsService authentication() {
        return new MapReactiveUserDetailsService(
                User
                        .withDefaultPasswordEncoder()
                        .username("jlong")
                        .password("pw")
                        .roles("USER")
                        .build()
        );
    }

    @Bean
    WebSocketHandlerAdapter webSocketHandlerAdapter() {
        return new WebSocketHandlerAdapter();
    }

    @Bean
    SimpleUrlHandlerMapping simpleUrlHandlerMapping(WebSocketHandler wsh) {
        return new SimpleUrlHandlerMapping(Map.of("/ws/greetings", wsh), 10);
    }

    @Bean
    WebSocketHandler webSocketHandler(GreetingService greetingService) {
        return session -> {
            var chat = session
                    .receive()
                    .map(WebSocketMessage::getPayloadAsText)
                    .map(GreetingRequest::new)
                    .flatMap(greetingService::greet)
                    .map(GreetingResponse::getMessage)
                    .map(session::textMessage);
            return session.send(chat);
        };
    }
}


@Controller
@RequiredArgsConstructor
class GreetingController {

    private final GreetingService greetingService;

    @MessageMapping("greetings")
    Flux<GreetingResponse> greet(@AuthenticationPrincipal Mono<UserDetails> principal) {
        return principal
                .map( UserDetails::getUsername)
                .map(GreetingRequest::new)
                .flatMapMany(greetingService::greet);
    }

}

@Service
class GreetingService {

    Flux<GreetingResponse> greet(GreetingRequest request) {
        return Flux
                .fromStream(Stream.generate(() -> new GreetingResponse("Hello " + request.getName() + " @ " + Instant.now())))
                .delayElements(Duration.ofSeconds(1));
    }
}

@Data
@AllArgsConstructor
@NoArgsConstructor
class GreetingRequest {
    private String name;
}

@Data
@AllArgsConstructor
@NoArgsConstructor
class GreetingResponse {
    private String message;
}


interface ReservationRepository extends ReactiveCrudRepository<Reservation, Integer> {
}

@Data
@AllArgsConstructor
@NoArgsConstructor
class Reservation {

    @Id
    private Integer id;
    private String name;
}

File: archived/service-kotlin/src/main/kotlin/com/example/service/ServiceApplication.kt
package com.example.service

import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.data.repository.reactive.ReactiveCrudRepository
import org.springframework.messaging.handler.annotation.MessageMapping
import org.springframework.security.config.Customizer
import org.springframework.security.config.annotation.rsocket.RSocketSecurity
import org.springframework.security.config.web.server.ServerHttpSecurity
import org.springframework.security.core.context.ReactiveSecurityContextHolder
import org.springframework.security.core.userdetails.MapReactiveUserDetailsService
import org.springframework.security.core.userdetails.User
import org.springframework.stereotype.Controller
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.server.ServerResponse
import org.springframework.web.reactive.function.server.body
import org.springframework.web.reactive.function.server.router
import org.springframework.web.reactive.handler.SimpleUrlHandlerMapping
import org.springframework.web.reactive.socket.WebSocketHandler
import org.springframework.web.reactive.socket.server.support.WebSocketHandlerAdapter
import reactor.core.publisher.Flux
import java.time.Duration
import java.time.Instant
import java.util.stream.Stream

@SpringBootApplication
class ServiceApplication


@Configuration
class SecurityConfiguration {

	private fun user(user: String) =
			User.withDefaultPasswordEncoder().username(user).password("pw").roles("USER").build()

	@Bean
	fun authentication() = MapReactiveUserDetailsService(user("rwinch"), user("jlong"))

	@Bean
	fun httpAuthorization(http: ServerHttpSecurity) = http
			.httpBasic(Customizer.withDefaults())
			.authorizeExchange { auth ->
				auth
						.pathMatchers("/reservations").authenticated()
						.anyExchange().permitAll()
			}
			.build()

	@Bean
	fun rsocketAuthorization(rsocketSecurity: RSocketSecurity) = rsocketSecurity
			.simpleAuthentication(Customizer.withDefaults())
			.authorizePayload { auth ->
				auth
						.route("greetings").authenticated()
						.anyExchange().permitAll()
			}
			.build()
}


@Configuration
class WebSocketConfiguration {

	@Bean
	fun wsha() = WebSocketHandlerAdapter()

	@Bean
	fun suhm(wsh: WebSocketHandler) = SimpleUrlHandlerMapping(mapOf("/ws/greetings" to wsh), 10)

	@Bean
	fun wsh(gs: GreetingService) = WebSocketHandler { session ->
		val chat = session.receive().map { it.payloadAsText }.map { GreetingRequest(it) }.flatMap { gs.greet(it) }.map { it.message }.map { session.textMessage(it) }
		session.send(chat)
	}
}

@Configuration
class HttpConfiguration {

	@Bean
	fun routes(rr: ReservationRepository) = router {
		GET("/reservations") {
			ServerResponse.ok().body(rr.findAll())
		}
	}

}


@Controller
class GreetingService {

	@MessageMapping("greetings")
	fun greet() = ReactiveSecurityContextHolder.getContext().map { it.authentication.name }.map { GreetingRequest(it) }.flatMapMany { greet(it) }

	fun greet(greetingRequest: GreetingRequest) =
			Flux
					.fromStream(Stream.generate {
						GreetingResponse("hello ${greetingRequest.name} @ ${Instant.now()}!")
					})
					.delayElements(Duration.ofSeconds(1))
}

data class Reservation(val id: Int, val name: String)
data class GreetingRequest(val name: String)
data class GreetingResponse(val message: String)

fun main(args: Array<String>) {
	runApplication<ServiceApplication>(*args)
}

interface ReservationRepository : ReactiveCrudRepository<Reservation, Int>
